x/stake/keeper.go

```diff
@@ -448,48 +432,132 @@ func (k Keeper) GetParams(ctx sdk.Context) (params Params) {
 		panic("Stored params should not have been nil")
 	}
 
-	err := k.cdc.UnmarshalBinary(b, &params)
-	if err != nil {
-		panic(err)
-	}
+	k.cdc.MustUnmarshalBinary(b, &params)
 	return
 }
 func (k Keeper) setParams(ctx sdk.Context, params Params) {
 	store := ctx.KVStore(k.storeKey)
-	b, err := k.cdc.MarshalBinary(params)
-	if err != nil {
-		panic(err)
-	}
+	b := k.cdc.MustMarshalBinary(params)
 	store.Set(ParamKey, b)
-	k.params = Params{} // clear the cache
+
+	// if max validator count changes, must recalculate validator set
+	if k.params.MaxValidators != params.MaxValidators {
+		k.updateValidators(ctx, store, sdk.Address{})
+	}
+	k.params = params // update the cache
```

Comment:
Caching considered harmful, see https://github.com/cosmos/cosmos-sdk/issues/957

Reply:
removed use of caches in the stake keeper